File: backend/src/Blindify.Domain/Enums/RaisonSignalement.cs
namespace Blindify.Domain.Enums;

/// <summary>Raison d'un signalement en direct (V2, section 12.4) — voir FlagEntryDto. Choisie par le
/// host ou un admin authentifié depuis l'écran de reveal, jamais par un joueur simple.</summary>
public enum RaisonSignalement
{
    /// <summary>Morceau qui n'a rien à faire dans le catalogue (trop obscur, inapproprié, doublon).</summary>
    PasSaPlace,

    /// <summary>Live, remix, reprise, mauvais match YouTube.</summary>
    MauvaiseVersion,

    /// <summary>Coupure, silence, volume, mauvaise qualité.</summary>
    AudioDefectueux,

    /// <summary>Titre, artiste, album ou année erronés.</summary>
    MetadonneesFausses,

    /// <summary>Ne correspond pas au thème (tags) de la série.</summary>
    HorsTheme,

    /// <summary>Track.RefrainStartMs à corriger.</summary>
    RefrainMalPlace,

    /// <summary>Commentaire obligatoire (voir GameHub.SignalerMorceau) — catégorie non prévue ci-dessus.</summary>
    Autre,
}

/// <summary>Classification fixe des raisons "bloquantes" (V2, section 12.4) — un morceau avec un
/// signalement bloquant non résolu est exclu de RoundService.SelectionnerMorceaux dès le prochain
/// ConfigurerPartie/RejouerPartie (voir GameConfig.ExclureMorceauxSignales). Une raison non bloquante
/// (thème, refrain, autre) n'empêche jamais le morceau d'être retiré — elle sert seulement à alimenter
/// le pipeline de curation (data/scripts/export_flags_csv.py).</summary>
public static class RaisonSignalementRules
{
    public static bool EstBloquante(RaisonSignalement raison) => raison switch
    {
        RaisonSignalement.PasSaPlace => true,
        RaisonSignalement.MauvaiseVersion => true,
        RaisonSignalement.AudioDefectueux => true,
        RaisonSignalement.MetadonneesFausses => true,
        _ => false,
    };
}

File: app/lib/models/raison_signalement.dart
/// Miroir de `Blindify.Domain.Enums.RaisonSignalement` (V2, section 12.4) — voir GameHub.SignalerMorceau.
enum RaisonSignalement {
  pasSaPlace,
  mauvaiseVersion,
  audioDefectueux,
  metadonneesFausses,
  horsTheme,
  refrainMalPlace,
  autre;

  /// Valeur envoyée au serveur — doit correspondre exactement au nom C# de l'enum (sérialisé en
  /// string côté contrat SignalR, voir ContractJsonOptions).
  String get code => switch (this) {
        RaisonSignalement.pasSaPlace => 'PasSaPlace',
        RaisonSignalement.mauvaiseVersion => 'MauvaiseVersion',
        RaisonSignalement.audioDefectueux => 'AudioDefectueux',
        RaisonSignalement.metadonneesFausses => 'MetadonneesFausses',
        RaisonSignalement.horsTheme => 'HorsTheme',
        RaisonSignalement.refrainMalPlace => 'RefrainMalPlace',
        RaisonSignalement.autre => 'Autre',
      };

  String get libelle => switch (this) {
        RaisonSignalement.pasSaPlace => "N'a rien à faire dans le catalogue",
        RaisonSignalement.mauvaiseVersion => 'Mauvaise version (live, remix, reprise...)',
        RaisonSignalement.audioDefectueux => 'Audio défectueux (coupure, volume, qualité...)',
        RaisonSignalement.metadonneesFausses => 'Métadonnées fausses (titre, artiste, année...)',
        RaisonSignalement.horsTheme => 'Hors thème de la série',
        RaisonSignalement.refrainMalPlace => 'Refrain mal placé',
        RaisonSignalement.autre => 'Autre',
      };

  /// Un commentaire est exigé côté UI pour cette raison — voir docs/architecture.md section 12.4.
  bool get commentaireObligatoire => this == RaisonSignalement.autre;
}
